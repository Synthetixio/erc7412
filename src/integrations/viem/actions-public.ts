import * as viem from 'viem'

import {
  prepareTransactionRequest as actionPrepareTransactionRequest,
  estimateContractGas as actionEstimateContractGas,
  simulateContract as actionSimulateContract
} from 'viem/actions'

import { callWithOffchainData } from '../..'
import type { OracleAdapter } from '../../types'

/**
 * Extend your viem client with the object returned by this function to automatically apply erc7412
 * required offchain data to your read calls
 */
export function createErc7412PublicActions (adapters: OracleAdapter[]) {
  return (client: viem.PublicClient) => {
    const actions = {
      call: async (args: viem.CallParameters): Promise<viem.CallReturnType> => {
        return {
          data: (
            await callWithOffchainData(
              [
                {
                  from: (args.account as any)?.address ?? viem.zeroAddress,
                  ...args
                }
              ],
              client,
              adapters
            )
          )[0]
        }
      },
      readContract: async (args: viem.ReadContractParameters): Promise<viem.ReadContractReturnType> => {
        return {
          data: viem.decodeFunctionResult({
            ...args,
            data: (
              await callWithOffchainData(
                [
                  {
                    from: (args.account as any)?.address ?? viem.zeroAddress,
                    data: viem.encodeFunctionData(args),
                    ...args
                  }
                ],
                client,
                adapters
              )
            )[0]
          })
        }
      },
      prepareTransactionRequest: async (args: viem.PrepareTransactionRequestParameters) => {
        try {
          return await actionPrepareTransactionRequest(client, args)
        } catch (err) {
          console.log('WARN: erc7412 not implemented for prepareTransactionRequest')
          throw err
        }
      },
      estimateContractGas: async (args: viem.EstimateContractGasParameters) => {
        try {
          return await actionEstimateContractGas(client, args)
        } catch (err) {
          console.log('WARN: erc7412 not implemented for estimateContractGas')
          throw err
        }
      },
      // TODO: types
      simulateContract: async (args: viem.SimulateContractParameters): Promise<any> => {
        try {
          return await actionSimulateContract(client, args)
        } catch (err) {
          const baseTxn = {
            from: (args.account as any)?.address ?? viem.zeroAddress,
            to: args.address,
            chain: args.chain,
            data: viem.encodeFunctionData(args),
            value: args.value
          }
          return {
            result: viem.decodeFunctionResult({
              ...args,
              data: (await callWithOffchainData([baseTxn], client, adapters))[0]
            }) as any
          }
        }
      },
      multicall: async (args: viem.MulticallParameters): Promise<viem.MulticallReturnType> => {
        if (args.contracts.length < 1) {
          throw new Error('must have at least one call for multicall')
        }

        const retvals = await callWithOffchainData(
          // todo: types have a problem with the fact it cannot verify that the array is at least 1 long
          args.contracts.map((c) => {
            return {
              from: c.address ?? viem.zeroAddress,
              data: viem.encodeFunctionData(c),
              ...c
            }
          }) as any,
          client,
          adapters
        )

        return retvals.map((r, i) => {
          return { result: viem.decodeFunctionResult({ ...args.contracts[i], data: r }), status: 'success' }
        })
      }
      // below functions are not included because they are not applicable
      // writeContract: async (args: viem.WriteContractParameters) => {},
    }

    return actions
  }
}
