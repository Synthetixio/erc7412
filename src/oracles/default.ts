import { type OracleAdapter } from '../types'
import type * as viem from 'viem'
import fetch from 'node-fetch'

export class DefaultAdapter implements OracleAdapter {
  constructor(private readonly oracleId: string, private readonly url: string) {}

  getOracleId(): string {
    return this.oracleId
  }

  async fetchOffchainData(
    _client: viem.Client,
    _oracleContract: viem.Address,
    oracleQuery: Array<{ query: viem.Hex; fee?: bigint }>
  ): Promise<Array<{ arg: viem.Hex; fee: bigint }>> {
    if (oracleQuery.length > 1) {
      throw new Error('only one query at a time is supported')
    }

    return [{ arg: await this.fetch(oracleQuery[0].query), fee: oracleQuery[0].fee ?? BigInt(0) }]
  }

  private async fetch(data: viem.Hex): Promise<viem.Hex> {
    const response = await fetch(this.url, {
      method: 'POST',
      headers: {
        'Content-Type': 'text/plain'
      },
      body: data
    })
    if (response.status !== 200) {
      throw new Error(`error fetching data (${response.status}): ${await response.text()}`)
    }
    return (await response.text()) as viem.Hex
  }
}
