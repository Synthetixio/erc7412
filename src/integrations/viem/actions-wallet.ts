import * as viem from 'viem'

import { prepareTransactionRequest as actionPrepareTransactionRequest } from 'viem/actions'

import { resolvePrependTransaction, makeTrustedForwarderMulticall, callWithOffchainData } from '../..'

import ITrustedMulticallForwarder from '../../../out/ITrustedMulticallForwarder.sol/ITrustedMulticallForwarder.json'

import type { TransactionRequest } from '../..'
import type { OracleAdapter } from '../../types'
import { getAccount } from './actions-public'

/**
 * Extend your viem client with the object returned by this function to automatically apply erc7412
 * required offchain data to your read calls
 */
export function createErc7412WalletActions (adapters: OracleAdapter[]) {
  return (client: viem.PublicClient) => {
    const actions = {
      prepareTransactionRequest: async (
        args: viem.PrepareTransactionRequestParameters
      ): Promise<viem.PrepareTransactionRequestReturnType> => {
        return await actions.prepareMulticallTransactionRequest({ txns: [args] })
      },

      prepareMulticallTransactionRequest: async (args: {
        txns: viem.PrepareTransactionRequestParameters[]
      }): Promise<viem.PrepareTransactionRequestReturnType> => {
        const prependedTxns: TransactionRequest[] = []

        if (args.txns.length < 1) {
          throw new Error('must have at least 1 transaction in multicall')
        }

        const payloadTxns: TransactionRequest[] = args.txns.map((t) => {
          const req: TransactionRequest = {
            ...t,
            from: getAccount(t.account)
          }

          return req
        })

        while (true) {
          const multicallTxn: viem.PrepareTransactionRequestParameters =
            payloadTxns.length > 1 || prependedTxns.length > 0
              ? {
                  ...makeTrustedForwarderMulticall([...prependedTxns, ...payloadTxns]),
                  account: args.txns[0].account,
                  chain: args.txns[0].chain
                }
              : args.txns[0]
          try {
            return await actionPrepareTransactionRequest(client, multicallTxn)
          } catch (err) {
            prependedTxns.push(await resolvePrependTransaction(err, client, adapters))
          }
        }
      },
      simulateContract: async (args: viem.SimulateContractParameters): Promise<viem.SimulateContractReturnType> => {
        const baseTxn = {
          from: (args.account as any)?.address ?? viem.zeroAddress,
          to: args.address,
          chain: args.chain,
          data: viem.encodeFunctionData(args),
          value: args.value
        }
        const preparedTxn = await actions.prepareTransactionRequest(baseTxn)
        return {
          request:
            preparedTxn.to === args.address
              ? args
              : {
                  ...preparedTxn,
                  abi: ITrustedMulticallForwarder.abi,
                  address: preparedTxn.to,
                  functionName: 'aggregate3Value',
                  args: viem.decodeFunctionData({ abi: ITrustedMulticallForwarder.abi, data: preparedTxn.data ?? '0x' })
                },
          result: viem.decodeFunctionResult({
            ...args,
            data: (await callWithOffchainData([baseTxn], client, adapters))[0]
          })
        } as any // TODO
      }

      // NOTE: cant override `sendRawTransaction` because the transaction is
      // already serialized and so its impossible to impart a new multicall
      // in front without breaking permissions
      // `sendTransaction` is not included because its behavior is not supposed to affect the transaction,
      // its just supposed to sign and send
    }

    return actions
  }
}
