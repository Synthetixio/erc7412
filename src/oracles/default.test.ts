import * as viem from 'viem'

import * as mod from './default'

import fetch from 'node-fetch'

jest.mock('node-fetch')

// prevent dumb bugs
;(BigInt.prototype as any).toJSON = function () {
  return this.toString()
}

describe('oracles/default.ts', () => {
  describe('DefaultAdapter', () => {
    const adapter = new mod.DefaultAdapter('DEFAULT', 'https://some-endpoint.com')

    describe('getOracleId()', () => {
      it('returns DEFAULT', () => {
        expect(adapter.getOracleId()).toEqual('DEFAULT')
      })
    })

    describe('fetchOffchainData()', () => {
      it('runs a query', async () => {
        const oracleQuery = [
          {
            query: viem.toHex('query one', { size: 32 }),
            fee: BigInt(10)
          }
        ]

        // eslint-disable-next-line
        jest.mocked(fetch).mockResolvedValue({ status: 200, text: async () => '0x2224' } as any)

        await expect(
          adapter.fetchOffchainData(
            viem.createClient({ transport: viem.custom({ request: async () => {} }) }),
            viem.zeroAddress,
            oracleQuery
          )
        ).resolves.toEqual([{ arg: '0x2224', fee: BigInt(10) }])

        expect(fetch).toHaveBeenCalledWith('https://some-endpoint.com', {
          method: 'POST',
          headers: {
            'Content-Type': 'text/plain'
          },
          body: viem.toHex('query one', { size: 32 })
        })

        await expect(
          adapter.fetchOffchainData(
            viem.createClient({ transport: viem.custom({ request: async () => {} }) }),
            viem.zeroAddress,
            [{ query: '0x' }]
          )
        ).resolves.toEqual([{ arg: '0x2224', fee: BigInt(0) }])
      })

      it('shows appropriate error when non 200 response comes back', async () => {
        const oracleQuery = [
          {
            query: viem.toHex('query one', { size: 32 }),
            fee: BigInt(10)
          }
        ]

        // eslint-disable-next-line
        jest.mocked(fetch).mockResolvedValue({ status: 404, text: async () => '0x2224' } as any)

        await expect(
          adapter.fetchOffchainData(
            viem.createClient({ transport: viem.custom({ request: async () => {} }) }),
            viem.zeroAddress,
            oracleQuery
          )
        ).rejects.toThrowErrorMatchingSnapshot()
      })

      it('supports only one query at a time', async () => {
        const oracleQuery = [
          {
            query: viem.toHex('query one', { size: 32 }),
            fee: BigInt(10)
          },
          {
            query: viem.toHex('query two', { size: 32 }),
            fee: BigInt(100)
          }
        ]

        await expect(
          adapter.fetchOffchainData(
            viem.createClient({ transport: viem.custom({ request: async () => {} }) }),
            viem.zeroAddress,
            oracleQuery
          )
        ).rejects.toThrow(new Error('only one query at a time is supported'))
      })
    })
  })
})
