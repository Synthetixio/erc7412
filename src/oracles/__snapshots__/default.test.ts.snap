// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oracles/default.ts DefaultAdapter fetchOffchainData() shows appropriate error when non 200 response comes back 1`] = `"error fetching data (404): 0x2224"`;
